src/jit/importer.cpp

```diff
@@ -1086,8 +1085,17 @@ GenTreePtr Compiler::impAssignStruct(GenTreePtr           dest,
     }
 
     assert(dest->gtOper == GT_LCL_VAR || dest->gtOper == GT_RETURN || dest->gtOper == GT_FIELD ||
-           dest->gtOper == GT_IND || dest->gtOper == GT_OBJ);
+           dest->gtOper == GT_IND || dest->gtOper == GT_OBJ || dest->gtOper == GT_INDEX);
 
+    if (dest->OperGet() == GT_LCL_VAR && src->OperGet() == GT_LCL_VAR &&
+        src->gtLclVarCommon.gtLclNum == dest->gtLclVarCommon.gtLclNum)
+    {
+        // Make this a NOP
+        return gtNewNothingNode();
+    }
+
+    // TODO-1stClassStructs: Avoid creating an address if it is not needed,
+    // or re-creating a Blk node if it is.
```

Comment:
In order to preserve existing behavior, it is sometimes necessary to get an address, and "sink" the struct object under an indirection, however it is often not needed downstream. Eliminating these early changes the IR, and thus the generated code (usually for the better, but surprisingly not always). I will expand this comment to make that a bit mroe clear.

---
In reply to: [77394689](https://github.com/dotnet/coreclr/pull/7027#discussion_r77394689) [](ancestors = 77394689)